cmake/modules/zephyr_default.cmake

```diff
@@ -81,9 +82,37 @@ include(${BOARD_DIR}/pre_dt_board.cmake OPTIONAL)
 # preprocess DT sources, and then, after we have finished processing
 # both DT and Kconfig we complete the target-specific configuration,
 # and possibly change the toolchain.
-include(generic_toolchain)
-include(dts)
-include(kconfig)
-include(soc)
-include(target_toolchain)
-include(kernel)
+list(APPEND zephyr_cmake_modules generic_toolchain)
+list(APPEND zephyr_cmake_modules dts)
+list(APPEND zephyr_cmake_modules kconfig)
+list(APPEND zephyr_cmake_modules soc)
+list(APPEND zephyr_cmake_modules target_toolchain)
+
+foreach(component ${SUB_COMPONENTS})
+  if(NOT ${component} IN_LIST zephyr_cmake_modules)
+    message(FATAL_ERROR
+      "Subcomponent '${component}' not default module for Zephyr CMake build system.\n"
+      "Please choose one or more valid components: ${zephyr_cmake_modules}"
+    )
+  endif()
+endforeach()
+
+foreach(module IN LISTS zephyr_cmake_modules)
+  # Ensures any module of type `${module}` are properly expanded to list before
+  # passed on the `include(${module})`.
+  # This is done twice to support cases where the content of `${module}` itself
+  # contains a variable, like `${BOARD_DIR}`.
+  string(CONFIGURE "${module}" module)
+  string(CONFIGURE "${module}" module)
+  include(${module})
+
+  list(REMOVE_ITEM SUB_COMPONENTS ${module})
+  if(DEFINED SUB_COMPONENTS AND NOT SUB_COMPONENTS)
+    # All requested Zephyr CMake modules have been loaded, so let's return.
+    return()
+  endif()
+endforeach()
+
+if(NOT DEFINED SUB_COMPONENTS)
+  include(kernel)
```

Comment:
shouldn't we unconditionally include kernel?

If SUB_COMPONENTS is defined the program will have returned before now.

Reply:
good point.

after the cleanup, this if can be removed.

will do.

Reply:
done